packages/dds/tree/src/test/openPolymorphism.spec.ts

```diff
@@ -0,0 +1,273 @@
+
+		/**
+		 * Item type registry.
+		 * @remarks
+		 * This doesn't have to be a mutable static.
+		 * For example libraries could export their implementations instead of adding them when imported,
+		 * then the top level code which pulls in all the libraries could aggregate the item types.
+		 *
+		 * TODO: document (and enforce/detect) when how late it is safe to modify array's used as allowed types.
+		 * These docs should ideally align with how late lazy type lambdas are evaluated (when the tree configuration is constructed, or an instance is made, which ever is first? Maybe define schema finalization?)
+		 */
+		const ItemTypes: ItemSchema[] = [];
+
+		// -------------
+		// Library using an Item
+
+		class Container extends sf.array("Container", ItemTypes) {}
+
+		// -------------
+		// Library defining an item
+
+		ItemTypes.push(TextItem);
+
+		// -------------
+		// Example use of container with generic code and down casting
+
+		const container = new Container();
+
+		// If we don't do anything special, the insertable type is never, so a cast is required to insert content.
+		container.insertAtStart(new TextItem({ text: "", location: { x: 0, y: 0 } }) as never);
+
+		// Items read from the container are typed as Item and have thew expected APIs:
+		const first = container[0];
+		first.foo();
+		first.location.x += 1;
+
+		// Down casting works as normal.
+		if (Tree.is(first, TextItem)) {
+			assert.equal(first.text, "foo");
+		}
+	});
+
+	it("mutable static registry, error cases", () => {
+		const ItemTypes: ItemSchema[] = [];
+		class Container extends sf.array("Container", ItemTypes) {}
+
+		// Not added to registry
+		// ItemTypes.push(TextItem);
+
+		const container = new Container();
+
+		// Should error due to out of schema content
+		assert.throws(
+			() =>
+				container.insertAtStart(new TextItem({ text: "", location: { x: 0, y: 0 } }) as never),
+			validateUsageError(/schema/),
+		);
+
+		// Modifying registration too late should error
+		assert.throws(() => ItemTypes.push(TextItem));
+	});
+
+	it("mutable static registry, recursive case", () => {
+		const ItemTypes: ItemSchema[] = [];
+
+		// Example recursive item implementation
+		class Container extends sf.array("Container", ItemTypes) {}
+		class ContainerItem extends sf.object("ContainerItem", {
+			...itemFields,
+			container: Container,
+		}) {
+			public static readonly description = "Text";
+			public static default(): TextItem {
+				return new TextItem({ text: "", location: { x: 0, y: 0 } });
+			}
+
+			public foo(): void {}
+		}
+
+		ItemTypes.push(ContainerItem);
+
+		const container = new Container();
+
+		container.insertAtStart(
+			new ContainerItem({ container: [], location: { x: 0, y: 0 } }) as never,
+		);
+	});
+
+	it("mutable static registry, safer editing API", () => {
+		const ItemTypes: ItemSchema[] = [];
+		class Container extends sf.object("Container", {
+			child: sf.optional(customizeSchemaTyping(ItemTypes).simplified<Item>()),
```

Comment:
This file shows several ways to do open polymorphism / schema child dependency injection. This one uses customizeSchemaTyping, and makes everything just work.